### PR TITLE
Secret lookup

### PR DESCRIPTION
**Pull request by chuckha**
79.5% coverage of key_pair.go. Can't test error cases without doing more dependency injection, but I think this is a nice compromise.

### CODE REVIEW

File: pkg/upgrade/key_pair.go [hunk 1]
```diff
@@ -8,42 +8,55 @@ import (
...
-	if config.ClusterField != "" {
-		return getEmbeddedKeyPair(cluster, config.ClusterField)
+	kc, ok := secret.Data[KubeconfigSecretKey]
+	if !ok {
+		return nil, errors.Errorf("item 'kubeconfig' not found in secret %q", name)
 	}
-
-	return k.getSecretRefKeyPair(cluster.Namespace, config.SecretRef)
+	out := make([]byte, base64.StdEncoding.DecodedLen(len(kc)))
+	n, err := base64.StdEncoding.Decode(out, kc)
```
Comment (chuckha) [inline pkg/upgrade/key_pair.go]: this actually doesn't need to be decoded, the client does that automatically which really confused the heck out of me.

File: pkg/upgrade/key_pair.go [hunk 2]
```diff
@@ -8,42 +8,55 @@ import (
...
+type keyPair struct {
 	Cert []byte `json:"cert"`
 	Key  []byte `json:"key"`
 }
 
-type keyPairGetter struct {
-	secretClient corev1client.SecretsGetter
+// secrets has already been scoped to namespace, so, .Secrets("my-namespace") returns one of these.
+type secrets interface {
+	Get(string, metav1.GetOptions) (*v1.Secret, error)
```
Comment (ncdc) [inline pkg/upgrade/key_pair.go]: If you wouldn't mind, please name the parameters

File: pkg/upgrade/key_pair.go (hunk 1 above)
Comment (ncdc) [reply pkg/upgrade/key_pair.go]: Does this mean you need to remove this `Decode` call?

File: pkg/upgrade/key_pair.go [hunk 3]
```diff
@@ -8,42 +8,55 @@ import (
...
+	out = out[:n]
+	return clientcmd.RESTConfigFromKubeConfig(out)
 }
 
-func getEmbeddedKeyPair(cluster *clusterapiv1alpha1.Cluster, path string) (*KeyPair, error) {
-	pathParts := strings.Split(path, ".")
+// NewRestConfigFromClusterField returns a rest.Config configured with the CA key pair found in the cluster's
+// object in the fieldpath specified. For example, "spec.providerSpec.value.caKeyPair" traverses the cluster
+// object going through each '.' delimited field.
+func NewRestConfigFromClusterField(cluster *clusterv1alpha1.Cluster, fieldPath, apiEndpoint string) (*rest.Config, error) {
```
Comment (ncdc) [inline pkg/upgrade/key_pair.go]: Should we call this `NewRestConfigFromCAClusterField`?

File: pkg/upgrade/key_pair.go [hunk 4]
```diff
@@ -8,42 +8,55 @@ import (
...
 	}
-
-	return k.getSecretRefKeyPair(cluster.Namespace, config.SecretRef)
+	out := make([]byte, base64.StdEncoding.DecodedLen(len(kc)))
+	n, err := base64.StdEncoding.Decode(out, kc)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	out = out[:n]
+	return clientcmd.RESTConfigFromKubeConfig(out)
```
Comment (ncdc) [inline pkg/upgrade/key_pair.go]: We should wrap the error here

File: pkg/upgrade/key_pair.go [hunk 5]
```diff
@@ -8,42 +8,55 @@ import (
...
 	"k8s.io/client-go/tools/clientcmd"
 	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services/certificates"
-	clusterapiv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	clusterv1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
 )
 
-// KeyPair contains a cert and key.
-type KeyPair struct {
+// KubeconfigSecretKey is the key where the kubeconfig is stored in the secret.
+const KubeconfigSecretKey = "kubeconfig"
```
Comment (ncdc) [inline pkg/upgrade/key_pair.go]: unexport?

File: pkg/upgrade/key_pair_test.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	secret := &secrets{
+		secret: &v1.Secret{
+			Data: map[string][]byte{
+				"cert": []byte(caCertificate),
+				"key":  []byte(caPrivateKey),
+			},
+		},
+	}
+	config, err := upgrade.NewRestConfigFromCASecretRef(secret, "name", "clustername", "https://example.com:6443")
+	if err != nil {
```
Comment (ncdc) [inline pkg/upgrade/key_pair_test.go]: Feel free to pull in testify and do `require.NoError(t, err)`, etc

Comment (chuckha) [reply pkg/upgrade/key_pair_test.go]: that seems like a big dependency for a very small syntax change. Is there any other reason you think we should be pulling that in?

File: pkg/upgrade/key_pair.go (hunk 1 above)
Comment (chuckha) [reply pkg/upgrade/key_pair.go]: yes, that is correct

File: pkg/upgrade/key_pair_test.go (hunk 6 above)
Comment (ncdc) [reply pkg/upgrade/key_pair_test.go]: I quite like testify for unit testing. If we pull it in, we can use it in all our tests.

Comment (chuckha) [reply pkg/upgrade/key_pair_test.go]: I don't think we need it for this statement as `err != nil` is an extremely common pattern in go. If it provides value in some other way in the future I'll be happy to pull it in

Comment (ncdc) [reply pkg/upgrade/key_pair_test.go]: Fair enough

File: pkg/upgrade/key_pair.go (hunk 5 above)
Comment (chuckha) [reply pkg/upgrade/key_pair.go]: yeah, that's fine. The purpose of this is to make it obvious what the secret key is and that purpose is still served with an unexported const 👍

File: pkg/upgrade/key_pair.go (hunk 2 above)
Comment (chuckha) [reply pkg/upgrade/key_pair.go]: yeah, no worries, that definitely improves readability

File: pkg/upgrade/key_pair.go (hunk 1 above)
Comment (chuckha) [reply pkg/upgrade/key_pair.go]: Actual result:

* Do not store secrets that you encode with base64. `kubectl` automatically encodes the data. If you *do* store b64 secrets, you will need to read them with `kubectl get secret ... | base64 --decode | base64 --decode`.
* Assume that providers are storing secrets that are *not* b64 encoded.

File: pkg/upgrade/key_pair.go (hunk 3 above)
Comment (chuckha) [reply pkg/upgrade/key_pair.go]: sure, seems a little more clear

**Comment by chuckha** [comment]
updated! ptal whenever you have time @ncdc

**Comment by ncdc** [comment]
LGTM!

**Pull request merged**